### PR TITLE
Make MAUI into 1 workload and lots of NuGets instead of a full workload

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change

This PR is to attempt to move towards a world where most MAUI bits are actually nugets. 

We may require this in order to work with .NET 7: 
 - https://github.com/dotnet/maui/pull/11665
 - https://github.com/dotnet/maui/pull/11946

Once this is merged you can still use all the same things as `<UseMaui>` will just now install nugets instead of use workloads. If you want to update the version of the nugets, you can just use the nuget package explorer - or use `<MauiVersion>`.

The only real fundamental change is that .NET MAUI is now entirely (almost) in NuGet packages instead of partially in nugets. This allows for cleaner updates. The aim of this PR is not to do away with workloads as they still offer several features we like, but rather move code that CHANGES into nugets. So we still want the magic of `<UseMaui>` or `<UseMauiCore>` but we also want to ensure that maui is determined by the nugets you install, not the workload you happen to be on.

Tasks:

- [x] Move all assemblies into nugets (core, controls essentials)
- [x] Move all targets into nugets (resizetizer and xamlc)
- [x] Make sure we still support net6
- [x] Write some docs on how things work

#### Testing

There are 2 ways to test this PR:

1. Remove all Microosft.Maui.* from your NuGet caches (or just clear it all)
2. Use the local .NET SDK for everything:  
   1. Run `dotnet cake --target=VS --pack --sln="/path/to/your/solution.sln"`
   3. The IDE will launch and you can run using F5
4. Use a global version of .NET MAUI _(only supported on 7.0.58 and later)_
   1. Download the NuGet artifacts from this PR or CI branch
   2. Extract the artifacts into a folder and add an entry in your NuGet.config to this folder (either in your solution directory or in VS)
   3. Open the project that you want to test with in the IDE
   5. Add the special property (just a temporary property until this PR is merged):
       ```xml
       <PropertyGroup>
           <UseMauiNuGets>true</UseMauiNuGets>
       </PropertyGroup>
       ```
   6. Add the new nugets to your project: 
       ```xml
       <ItemGroup>
           <PackageReference Include="Microsoft.Maui.Controls" Version="VERSION_FROM_CI_HERE" />
       </ItemGroup>
       ```
   7. Due to https://github.com/microsoft/WindowsAppSDK/issues/3251 you will also need to add the `Microsoft.Windows.SDK.BuildTools` package to windows:
       ```xml
       <ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">
           <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.755" />
       </ItemGroup>
       ```
   8. Run using F5

### CODE REVIEW

File: src/Workload/README.md
Comment (jonathanpeppers) [inline src/Workload/README.md:23]: Can you update this doc to describe the layout of the workload vs what is a NuGet?

File: src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets
Comment (jonathanpeppers) [inline src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets:6]: Do we lose the `UseMaui` capability? Lutz heavily uses this with VS's telemetry.

`UseMauiCore` is also nice if someone wants to find out usage about that.

File: src/Workload/Microsoft.Maui.Sdk/Sdk/Microsoft.Maui.Sdk.After.targets
Comment (jonathanpeppers) [inline src/Workload/Microsoft.Maui.Sdk/Sdk/Microsoft.Maui.Sdk.After.targets:16]: These need to go in `WorkloadManifest.targets`, so that Lutz can get telemetry for customers opening MAUI projects and they have no MAUI workloads installed.

**Review by jonathanpeppers — COMMENTED** [review]
General questions:
* Nuke your NuGet cache and...
* What is the time of `dotnet restore -bl`, share a `.binlog`?
* Does the IDE "act funny" while restore is going on?

File: src/Workload/Microsoft.Maui.Sdk/Sdk/Microsoft.Maui.Sdk.After.targets
Comment (mattleibow) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/Microsoft.Maui.Sdk.After.targets:16]: Gotcha! Good catch.

File: src/Workload/README.md
Comment (mattleibow) [reply src/Workload/README.md:23]: Busy on that, but yeah. Gonna be epic and all that. 😄

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (jonathanpeppers) [inline src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:65]: What does a customer's `.csproj` look like if they want to try a nightly build of MAUI? Will they have to update all of these packages?

Should there be a single `Microsoft.Maui` meta package?

We might also test that you can downgrade to older versions -- the way this looks (I think?) NuGet will just pick the higher one if you try to put a version older than what the workload says.

Comment (mattleibow) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:65]: I created this meta package `Microsoft.Maui.Controls` so I probably need to clean this up now.

**Comment by yevgeny-sotnikov** [comment]
@davidortinau 

This code will allow you to split the Maui monolith into subprojects, which will allow you to collect separate parts of the functionality without pulling unnecessary dependencies. This will allow new versions of the framework to be released on a daily basis. It will also simplify the process of creating and solving issues, which will increase the number of contributors

This PR is extremely important for the future of .NET mobile community.

Thank you

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (mattleibow) [reply src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets]: Done! All nice and clean and only importing the top-level packages.

File: src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets
Comment (mattleibow) [inline src/Workload/Microsoft.Maui.Sdk/Sdk/BundledVersions.in.targets:23]: @Redth  Should these still be PrivateAssets=all? Right now, UseMaui in a class library does not add any additional dependencies on the final nupkg, but that may have other issues. If we remove this, then the final nupkg will have all these dependencies.

The downside is that if we are using UseMaui in other repos and are building against preview versions of things that are not public, those versions will now be hardcoded in their final packages. For example, I sometimes use CI builds of maui to build SkiaSharp.Views.Maui and then publish ahead of time. Removing the PrivateAssets=all will now mean I have to wait for things to go public.

File: src/Workload/README.md
Comment (mattleibow) [reply src/Workload/README.md]: https://github.com/dotnet/maui/blob/dev/testing-nugets/docs/design/NuGets.md

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props
Comment (jonathanpeppers) [inline src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props:2]: So one thing I remember from early days... If we move these out of a `AutoImport.props` that lives with the workload, imagine the behavior:

1. I install Visual Studio for the first time & the MAUI (VS) workload
2. File -> New MAUI project
3. NuGet does it's thing trying to restore

Let's imagine the NuGet restore takes multiple seconds -- as it likely will the first time.

So there will be a moment where your project/solution tree looks completely empty. Then NuGet restores, you get a `Microsoft.Maui.Controls.Globs.props` and magically all the files appear in your solution tree.

Do weird things happen here in the IDE? Can we test in a VM?

File: src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets
Comment (jonathanpeppers) [inline src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets:9]: Might read better with the following line to do:
```suggestion
  <ItemGroup>
    <ProjectCapability  Condition=" '$(UseMaui)' == 'true' or '$(UseMauiCore)' == 'true' " Include="UseMauiCore" />
```

File: .gitignore
Comment (mattleibow) [reply .gitignore:6]: The whole folder is removed and now the AOT profiles now live in the correct location in the Build Tasks project - along with all the targets and props.

File: src/Workload/README.md
Comment (mattleibow) [reply src/Workload/README.md:113]: Not sure if there is a new recommendation just yet, but this property still works.

Comment (jonathanpeppers) [reply src/Workload/README.md:113]: Don't we need to understand how customers can more easily upgrade/downgrade projects? Is that the main goal of this work?

My fear is that customers won't actually be able to downgrade, and NuGet will always pick the higher one (if workload is higher than .csproj).

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/net6.0-ios10.0/Microsoft.Maui.Controls.Build.Tasks.props
Comment (mattleibow) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/net6.0-ios10.0/Microsoft.Maui.Controls.Build.Tasks.props:1]: We can do something, this was just the fastest. It should not matter the version here as the normal nuget fallback version will kick in. I was testing this with net7. However, we could rename to say `ios` and then use an include/link to include the `ios` folder as `$(_netversion)-ios` I'll have a look at that as it keeps the codebase cleaner...

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props
Comment (mattleibow) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props:2]: We will test this but what I noticed that the default project includes everything and then we go around removing and updating. Everything that is included by default is/should already be in the project with potentially incorrect build actions. Then this nuget/the workload would add items to new groups.

However, this file is still up for grabs and can be moved back if there are any issues.

File: src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets
Comment (mattleibow) [reply src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets:9]: This was the original version - the merge is just bad:

https://github.com/dotnet/maui/commit/cca129b12482ba1b969c2d785a4629e23b126044#diff-3042f8c369245b07c4cbf2ce9ad84538a813ee7483e623863896611dc96f4edeR2-R5

This got me wondering if it was wrong and UseMauiCore should only be set in certain situations? I'll check with lutz.

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/net6.0-ios10.0/Microsoft.Maui.Controls.Build.Tasks.props
Comment (mattleibow) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/net6.0-ios10.0/Microsoft.Maui.Controls.Build.Tasks.props:1]: This is going to be a bit harder since there are 2 values that need to change: .NET version and OS verion. Not sure I can have a net12.0-ios10 if the net12 has a min version of the OS ios20...

File: src/Workload/README.md
Comment (mattleibow) [reply src/Workload/README.md:113]: I fixed this to only include "missing" nugets. I can detect that Maui Controls is in the csproj, so don't add. This way, the csproj _always_ wins. And I also have reduced the nugets to just be the single, top-level package. And in most cases, this is just `Microsoft.Maui.Controls`.

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/net6.0-ios10.0/Microsoft.Maui.Controls.Build.Tasks.props
Comment (mattleibow) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/net6.0-ios10.0/Microsoft.Maui.Controls.Build.Tasks.props:1]: Going to leave this for now, and then in a new PR try and figure out some build logic for this. We need a version in the nuget package so it can install so this may mean seeing what the min supported version and things are.

File: docs/design/NuGets.md
Comment (mattleibow) [reply docs/design/NuGets.md:158]: Let me test and see. All the nugets should go in the same place - just like before with blazor and graphics. We just added a few more packages.

Comment (mattleibow) [reply docs/design/NuGets.md:158]: After a check and things, the only part that uses the workload packs are the templates step, and that sets up its own nuget context and folders. SO this should not be an issue.

File: src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets
Comment (jonathanpeppers) [inline src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets:32]: This looks like it only goes 1 version behind. Does anything support back to `net6.0` projects?

I kept `net6.0` support in the Android workload, because I didn't know yet what the official support is.

**Review by jonathanpeppers — COMMENTED** [review]
Does anything test that `net7.0` projects continue to work? What about `net6.0`?

File: src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets
Comment (mattleibow) [reply src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets:32]: We have tests on CI that test all this on both net6 and net7. All our tests are duplicated for the supported TFMs.

Comment (jonathanpeppers) [reply src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets:32]: So if I look at the workload manifest:

https://github.com/dotnet/maui/blob/ea8c7fad3a0a33b41744506757fe645a260028b4/src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.json

It has:

* Microsoft.Maui.Core (has the .NET 8 stuff?)
* Microsoft.Maui.Sdk.net6
* Microsoft.Maui.Sdk.net7

What are the values for `@MAUI_DOTNET_VERSION_MAJOR@` and `@MAUI_PREVIOUS_DOTNET_RELEASED_NUGET_VERSION@`? By their naming it sounds like these are 8 and 7.

File: src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.json
Comment (jonathanpeppers) [inline src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.json:71]: I think you can remove these 3 packs from all the `maui-*` platform workloads, because they all extend `maui-core`?

It wouldn't cause an isuse, just makes things cleaner.

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props
Comment (mattleibow) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props:2]: Doing several tests and this seems OK. We no longer add/remove things during the evaluation. We do it before the core build.

File: src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets
Comment (mattleibow) [reply src/Workload/Microsoft.NET.Sdk.Maui/WorkloadManifest.in.targets:32]: If we decide to backport this to net7 servicing, then this will involve more conditions and tests in the targets file to either install nugets for later net7 or workloads. So maybe we will have a condition that is `currentVersion = previousTFM AND currentVersion > switchVersion`

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props
Comment (mattleibow) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props:2]: Further testing seems to have show _why_ the IDE was slow before... If I set the project capabilities in the main Sdk.targets (or any included directly), the IDE first goes to some sort of generic "Start" run button and then after all the packages are restored select the correct windows machine.

However, if I put the project capabilities in some "after" targets, then it works. I am not sure why this is the case as they technically should be all fine in any order (I will try look at msbuild and the main sdk) but if they are imported last, then it works:

```xml
  <PropertyGroup>
    <AfterMicrosoftNETSdkTargets>$(AfterMicrosoftNETSdkTargets);$(MSBuildThisFileDirectory)Microsoft.Maui.Sdk.After.targets</AfterMicrosoftNETSdkTargets>
  </PropertyGroup>
```

**Comment by mattleibow** [comment]
Thanks for all the reviews @jonathanpeppers ! I did observe the cray things again with the IDE and the capabilities. I think I found out where/what caused the issues, but we are safe in this PR. I can reproduce and then fix the issue reliably, so I am almost confident it is not going to cause issues...

**Comment by yunusefendi52** [comment]
I have a couple questions for this:
1. Can I use MAUI without downloading the workload? like the way Xamarin.Forms used to
2. Is this also going to make MAUI supports _any_ underlying dotnet ios/android version? Current version (7.0.59) needs dotnet ios 16.2 but if I use dotnet ios 16.0 it wouldn't work because MAUI needs 16.2.
Xamarin.Forms doesn't really care what Xamarin ios/android version you use technically, if MAUI doesn't support that then I think it's still a stepback. I'd say dotnet workload system _is_ a stepback generally

**Comment by mattleibow** [comment]
@yunusefendi52 to answer your question, almost.

**TL;DR**

No, we still need workloads. This is phase 1 of reducing workload version restrictions. This PR makes it so you don't need to care what version you have installed as you now use nugets for that.

**Long answer**

This PR does not aim to make it workload-free as we do get some nice things. However, Windows and the IDE need more attention. One of the core issues is that the IDE decides what menu options and toolbar actions to do based on project capabilities - and these cannot be in nugets.

In the old world, Xamarin.Forms had separate projects with project type GUIDs to do this. In .NET 6+ this is now the responsibility of the workloads. This allows actions to be influenced by a SDK outside the IDE.

During the hacks I was doing recently, it appears that because workloads are just a "wrapper" around the existing SDK importing and things like that - think the MSBuild.SDK.Extras if you are familiar, we can potentially have our own SDK and then will not need workloads.

But, the core goal of this PR is to get all changing logic into nugets, so _almost_ we don't _need_ workloads. What this presents itself after the merge of this PR is that it won't matter what workload you have installed because there is no changing code in there. If you want a different maui version, either use MauiVrsion - or just select the version using the NuGet package manager.

**Review by Eilon — APPROVED** [review]
Here's what I've done:

* Run the branch locally on a Windows machine to run the Windows and Android samples in the MAUI solution
* On the same machine, I created new MAUI / MAUI Blazor projects and run them with this change
* Did editing in VS, debugging, etc.
* Reviewed the code changes in the PR
* Inspected the contents of the built NUPKGs

I do have two areas of unknowns/concerns that I think need to be addressed at some point:

1. There is a very long VS hang after debugging the Controls Sample app on Windows in the MAUI solution. On a very high end dev machine VS hangs for 2-3 minutes and the VS UI is "ghosted" (unresponsive, not rendering). It does eventually wake up. This doesn't happen with Android. Running without debugging is not affected. This has been reported to VS. I think this could make working in the MAUI solution very difficult. I'm not sure if this affects less complex solutions.
2. While I verified that the built NuGet packages indeed contain what I expected them to contain, I am not able to verify whether this change would in fact enable customers to pick their MAUI versions, do various side-by-side scenarios, etc.

In totality, this PR looks good, so I sign off.

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props
Comment (jonathanpeppers) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props:2]: Did we resolve this one? I wonder if we're going to have to keep `AutoImport.props` in the workload.

File: src/Core/src/nuget/buildTransitive/Microsoft.Maui.Core.props
Comment (jonathanpeppers) [inline src/Core/src/nuget/buildTransitive/Microsoft.Maui.Core.props:1]: Can we delete this file?

**Review by jonathanpeppers — COMMENTED** [review]
I'll finish reviewing next week. Mainly just want to read the files GitHub says changed since last time.

Is there any concern to merge this prior to branching .NET 8 Preview 1? Maybe we should merge right after branching is done?

File: src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props
Comment (mattleibow) [reply src/Controls/src/Build.Tasks/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Controls.Globs.props:2]: I think we did. As far as I can tell - and probably also @Eilon - things seem to work. I am keeping the file just in case the things explode. If there are any issues, we can add the bits (or all) back.

Hopefully we can get some testing in net 8 previews and then decide to backport (or not).

**Pull request merged**
